### DISCUSSION TITLE
Relating the PointNet++ classification model to the original paper architecture description.

### DISCUSSION THREAD

**Discussion by DanielTakeshi** [pyg-team/pytorch_geometric#3924] [Q&A] [answered]
Hello!

Thank you so much for this great resource. I have been looking at the PointNet++ implementation here:

https://github.com/pyg-team/pytorch_geometric/blob/master/examples/pointnet2_classification.py
https://github.com/pyg-team/pytorch_geometric/blob/master/examples/pointnet2_segmentation.py

as well as the Google Colab tutorial:  https://github.com/pyg-team/pytorch_geometric/issues/141

I now feel much more comfortable understanding PointNet++. 

<hr>

Right now I'm trying to figure out how the description from the NeurIPS 2017 paper (https://arxiv.org/abs/1706.02413) translates into the PyTorch code. For simplicity, let's just consider the **classification** version of PointNet++, and in the code it is here:

https://github.com/pyg-team/pytorch_geometric/blob/50b7bfc4a59b5b6f7ec547ff862985f3b2e22798/examples/pointnet2_classification.py#L43-L61

The paper's appendix says this regarding the classification architecture:

![Screenshot from 2022-01-22 10-51-31](https://user-images.githubusercontent.com/5532096/150645740-dfe33f55-a771-4028-8f9f-11534a932ea9.png)

From the paper, it says that the notation of `SA(K, r, [...])` is meant to be a set abstraction level with `K` local regions of ball radius `r` (unfortunately, `K` is overloaded later as the number of classes...). In the code, I am trying to figure out where this `K` parameter (the number of local regions) is located. I'm  not sure if I see it, because the `SAModule` seems to instead use a `self.ratio` term:

https://github.com/pyg-team/pytorch_geometric/blob/50b7bfc4a59b5b6f7ec547ff862985f3b2e22798/examples/pointnet2_classification.py#L12-L27

Is this an equivalently way of describing `K` local regions? I am unsure, especially because this "ratio" seems to be a sampling ratio in the farthest point algorithm:

https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html?highlight=fps#torch_geometric.nn.pool.fps

which seems a bit different from the number of local regions? Though I guess increasing the number of local regions could be like increasing the sampling ratio?

For the most part, the other parts of the code seem to match the paper:

 - The 0.2 and 0.4 in the first two set abstraction levels are exactly as described in the paper.
 - There are two SA layers, and then a third SA layer which is implemented using `GlobalSAModule` instead of `SAModule`
 - The MLPs seem to match what's in the paper.

Hence why I'm curious about this missing parameter.

**Comment by rusty1s** [comment, marked as answer]
Instead of fixing the number of regions `K`, we use the `ratio` argument instead, which denotes the fraction of regions we want to use in relation to the number of input points. If you have 1024 points in your cloud, and set `K=512`, then this is equivalent to setting `ratio=0.5`.

**Reply by DanielTakeshi** [reply]
Thanks @rusty1s for your clarification!
